src/ledger/tests.zig

```diff
@@ -44,3 +244,175 @@ pub fn freshDir(path: []const u8) !void {
+        defer allocator.free(chunk);
+        try entries.append(try sig.bincode.readFromSlice(allocator, Entry, chunk, .{}));
+    }
+    return entries.toOwnedSlice();
+}
+
+const State = TestState("global");
+const DB = TestDB("global");
+
+pub fn TestState(scope: []const u8) type {
+    return struct {
+        db: BlockstoreDB,
+        registry: sig.prometheus.Registry(.{}),
+        lowest_cleanup_slot: sig.sync.RwMux(Slot),
+        max_root: std.atomic.Value(Slot),
+
+        // if this leaks, you forgot to call `TestState.deinit`
+        _leak_check: *u8,
+
+        /// This is used instead of std.testing.allocator because it includes more stack trace frames
+        /// std.testing.allocator is already the same exact allocator, just with a call to detectLeaks
+        /// run at the end of the test. TestState does the same, so we can use the gpa directly.
+        var gpa = std.heap.GeneralPurposeAllocator(.{ .stack_trace_frames = 100 }){};
+        /// This is private to ensure _leak_check is initialized before this is used.
+        const _allocator = gpa.allocator();
+
+        const Self = @This();
+
+        pub fn init(comptime test_name: []const u8) !*Self {
+            const self = try _allocator.create(Self);
+            self.* = .{
+                .db = try TestDB(scope).initCustom(_allocator, test_name),
+                .registry = sig.prometheus.Registry(.{}).init(_allocator),
+                .lowest_cleanup_slot = sig.sync.RwMux(Slot).init(0),
+                .max_root = std.atomic.Value(Slot).init(0),
+                ._leak_check = try std.testing.allocator.create(u8),
+            };
+            return self;
+        }
+
+        pub fn allocator(_: Self) Allocator {
+            return _allocator;
+        }
+
+        pub fn shredInserter(self: *Self) !ledger.ShredInserter {
+            return ledger.ShredInserter.init(_allocator, test_logger, &self.registry, self.db);
+        }
+
+        pub fn writer(self: *Self) !ledger.BlockstoreWriter {
+            return try ledger.BlockstoreWriter.init(
+                _allocator,
+                test_logger,
+                self.db,
+                &self.registry,
+                &self.lowest_cleanup_slot,
+                &self.max_root,
+            );
+        }
+
+        pub fn reader(self: *Self) !ledger.BlockstoreReader {
+            return try ledger.BlockstoreReader.init(
+                _allocator,
+                test_logger,
+                self.db,
+                &self.registry,
+                &self.lowest_cleanup_slot,
+                &self.max_root,
+            );
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.db.deinit();
+            self.registry.deinit();
+            std.testing.allocator.destroy(self._leak_check);
+            _allocator.destroy(self);
+            _ = gpa.detectLeaks();
+        }
+    };
+}
+
+pub fn TestDB(scope: []const u8) type {
```

Comment:
why not just have `scope` as a parameter to `init` ? or have the scope prefix as a global constant in the file so you have `.init(TEST_DB_SCOPE + "blockstoreTest")` - want to minimize the layers of abstraction wherever possible imo

Reply:
ok that makes sense - thanks for explaining - we can keep it as is